include/rocksdb/trace_record.h

```diff
@@ -0,0 +1,133 @@
+//  Copyright (c) 2011-present, Facebook, Inc.  All rights reserved.
+//  This source code is licensed under both the GPLv2 (found in the
+//  COPYING file in the root directory) and Apache 2.0 License
+//  (found in the LICENSE.Apache file in the root directory).
+
+#pragma once
+
+// To do: ROCKSDB_LITE ?
+
+#include "rocksdb/db.h"
+#include "rocksdb/slice.h"
+#include "rocksdb/write_batch.h"
+
+namespace ROCKSDB_NAMESPACE {
+
+// Supported Trace types.
+enum TraceType : char {
+  kTraceNone = 0,
+  kTraceBegin = 1,
+  kTraceEnd = 2,
+  kTraceWrite = 3,
+  kTraceGet = 4,
+  kTraceIteratorSeek = 5,
+  kTraceIteratorSeekForPrev = 6,
+  // Block cache related types.
+  kBlockTraceIndexBlock = 7,
+  kBlockTraceFilterBlock = 8,
+  kBlockTraceDataBlock = 9,
+  kBlockTraceUncompressionDictBlock = 10,
+  kBlockTraceRangeDeletionBlock = 11,
+  // For IOTracing.
+  kIOTracer = 12,
+  // For query tracing
+  kTraceMultiGet = 13,
+  // All trace types should be added before kTraceMax
+  kTraceMax,
+};
+
+// Base class for all types of trace records.
+class TraceRecord {
+ public:
+  explicit TraceRecord(uint64_t ts = 0) : timestamp(ts) {}
+  virtual ~TraceRecord() {}
+
+  virtual TraceType GetType() const = 0;
+
+  // Timestamp (in microseconds) of this trace.
+  uint64_t timestamp;
+};
+
+// Base class for all query types of trace records.
+class QueryTraceRecord : public TraceRecord {
```

Comment:
Is it better to add the trace type as the member?

Reply:
It's inherited from the base class `TraceRecord::GetType()`.

Reply:
Trace type of a trace record should be read only, thus not class member.